pkg/controller/generic_control.go

```diff
@@ -0,0 +1,256 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/pingcap/tidb-operator/pkg/scheme"
+	corev1 "k8s.io/api/core/v1"
+	apiequality "k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// GenericControlInterface is a wrapper to manage typed object that managed by an arbitrary controller
+type TypedControlInterface interface {
+	CreateOrUpdateConfigMap(controller runtime.Object, cm *corev1.ConfigMap) (*corev1.ConfigMap, error)
+	Delete(controller, obj runtime.Object) error
+}
+
+type typedWrapper struct {
+	GenericControlInterface
+}
+
+// NewTypedControl wraps a GenericControlInterface to a TypedControlInterface
+func NewTypedControl(control GenericControlInterface) TypedControlInterface {
+	return &typedWrapper{control}
+}
+
+func (w *typedWrapper) Delete(controller, obj runtime.Object) error {
+	return w.GenericControlInterface.Delete(controller, obj)
+}
+
+func (w *typedWrapper) CreateOrUpdateConfigMap(controller runtime.Object, cm *corev1.ConfigMap) (*corev1.ConfigMap, error) {
+	result, err := w.GenericControlInterface.CreateOrUpdate(controller, cm, MergeConfigMap)
+	if err != nil {
+		return nil, err
+	}
+	return result.(*corev1.ConfigMap), nil
+}
+
+// GenericControlInterface manages generic object that managed by an arbitrary controller
+type GenericControlInterface interface {
+	CreateOrUpdate(controller, obj runtime.Object, mergeFn MergeFn) (runtime.Object, error)
+	Delete(controller, obj runtime.Object) error
+}
+
+// MergeFn knows how to merge a desired object into the current object.
+// Typically, merge should only set the specific fields the caller wants to control to the existing object
+// instead of override a whole struct. e.g.
+//
+// Prefer:
+//     existing.spec.type = desired.spec.type
+//     existing.spec.externalTrafficPolicy = desired.spec.externalTrafficPolicy
+// Instead of:
+//     existing.spec = desired.spec
+//
+// However, this could be tedious for large object if the caller want to control lots of the fields,
+// if there is no one else will mutate this object or cooperation is not needed, it is okay to do aggressive
+// override. Note that aggressive override usually causes unnecessary updates because the object will be mutated
+// after POST/PUT to api-server (e.g. Defaulting), an annotation based technique could be used to avoid such
+// updating: set a last-applied-config annotation and diff the annotation instead of the real spec.
+type MergeFn func(existing, desired runtime.Object) error
```

Comment:
`server-side apply` is a better solution in the future.

ref: https://kubernetes.io/docs/reference/using-api/api-concepts/#server-side-apply